http2: make maximum invalid frames and maximum rejected streams configurable

lundibundi opened the pull request:
**http2: replace direct array usage with struct for js_fields_**

**http2: allow to configure maximum tolerated invalid frames**

**test: update and harden http2-reset-flood**
* use new maxSessionInvalidFrames to lower the needed frames
* slow down requests to generate less redundant after-session-close
  requests

**http2: make maximum tolerated rejected streams configurable**
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
Closes: https://github.com/nodejs/node/issues/30505
This should also decrease the flakiness of `test-http2-reset-flood`.
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Also, do we have a dedicated test for argument validation of `createServer`? I'm not sure where to put them.

/cc @nodejs/http2

ZYSzys commented on doc/api/http2.md:
```diff
@@ -1941,6 +1941,9 @@ error will be thrown.
 added: v8.4.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/REPLACEME
```
```suggestion
    pr-url: https://github.com/nodejs/node/pull/30534
```

lundibundi replied:
Done.

ZYSzys commented on doc/api/http2.md:
```diff
@@ -2053,6 +2065,9 @@ server.listen(80);
 added: v8.4.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/REPLACEME
```
```suggestion
    pr-url: https://github.com/nodejs/node/pull/30534
```

lundibundi replied:
Done.

ZYSzys commented:
> Also, do we have a dedicated test for argument validation of `createServer`? I'm not sure where to put them.

Seem like we actually don't have a dedicated test for argument validation of `createServer`, but we test things like this by adding a new test file called `test-http2-too-many-*.js`.

Maybe add a new test and name it as `test-http2-too-many-invalid-frame.js` is suitable here.

addaleax commented on lib/internal/http2/core.js:
```diff
@@ -937,6 +944,24 @@ function setupHandle(socket, type, options) {
+      kSessionMaxInvalidFrames,
+      options.maxSessionInvalidFrames,
+      !isBigEndian
+    );
```
I think something like 

```js
const uint32 = new Uint32Array(this[kNativeFields].buffer, kMaxInvalidFrames, 1);
uint32[0] = options.maxSessionInvalidFrames;
```

Basically do the same thing without having to worry about endianness?

lundibundi commented:
Added dedicated test for `maxSessionInvalidFrames` that should be more reliable.
Added simple options validation testing for `createServer`/`createSecureServer`.
Fixed missing `changelog` for `maxSessionRejectedStreams` option.

lundibundi commented:
(nothing changed, force pushed to resolve conflicts)

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -1941,6 +1941,12 @@ error will be thrown.
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/30534
+    description: Added maxSessionRejectedStreams option with a default of 100.
```
```suggestion
    description: Added `maxSessionRejectedStreams` option with a default of
                 `100`.
```

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -1941,6 +1941,12 @@ error will be thrown.
+    description: Added maxSessionRejectedStreams option with a default of 100.
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/30534
+    description: Added maxSessionInvalidFrames option with a default of 1000.
```
```suggestion
    description: Added `maxSessionInvalidFrames` option with a default of
                 `1000`.
```

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -2053,6 +2068,12 @@ server.listen(80);
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/30534
+    description: Added maxSessionRejectedStreams option with a default of 100.
```
```suggestion
    description: Added `maxSessionRejectedStreams` option with a default of
                 `100`.
```

vsemozhetbyt commented on doc/api/http2.md:
```diff
@@ -2053,6 +2068,12 @@ server.listen(80);
+    description: Added maxSessionRejectedStreams option with a default of 100.
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/30534
+    description: Added maxSessionInvalidFrames option with a default of 1000.
```
```suggestion
    description: Added `maxSessionInvalidFrames` option with a default of
                 `1000`.
```

lundibundi commented:
@vsemozhetbyt done, thanks.

addaleax commented:
Landed in 10f5fa75136b...988034be6aece7bd103249e18ac7f0a5d3f92545 :tada:
